verify on chain config

FelixFan1992 opened the pull request.

jmank88 commented on core/scripts/chaincli/handler/verify_config.go:
```diff
@@ -0,0 +1,490 @@
+	ocr2config "github.com/smartcontractkit/libocr/offchainreporting2/confighelper"
+	ocr2types "github.com/smartcontractkit/libocr/offchainreporting2/types"
+
+	"github.com/smartcontractkit/chainlink/core/cmd"
+	registry20 "github.com/smartcontractkit/chainlink/core/gethwrappers/generated/keeper_registry_wrapper2_0"
+	"github.com/smartcontractkit/chainlink/core/logger"
```
```suggestion
	"github.com/smartcontractkit/chainlink/v2/core/cmd"
	registry20 "github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/keeper_registry_wrapper2_0"
	"github.com/smartcontractkit/chainlink/v2/core/logger"
```

ferglor commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,307 @@
+		diffs := 0
+
+		log.Infof("start comparing data for node %s", nodeAddr)
+		if node.CSAKeys[0].NodeAddress != ni.CSAKeys[0].NodeAddress {
```
Does it make sense to implement a single `Equals` function on the `NodeInfo` type that just returns a boolean, and then move all of these comparisons into that function?

FelixFan1992 replied:
done

FelixFan1992 commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,307 @@
+func fetchRawBytes(ctx context.Context, url string, log logger.Logger) []byte {
+	client := http.DefaultClient
+
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
```
done

FelixFan1992 commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,307 @@
+	if !h.cfg.VerifyNodes {
+		return
+	}
+	nodeInfos := fetchNodeInfos(ctx, h.cfg.NodeConfigURL, log)
```
done

FelixFan1992 commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,307 @@
+	oracleIdentities := make([]ocr2config.OracleIdentityExtra, len(cls))
+	for i, cl := range cls {
+		wg.Add(1)
+		go func(i int, cl cmd.HTTPClient) {
```
done

FelixFan1992 commented:
1. stopped using `reflect.DeepEqual` due to its drawbacks
2. removed lots of dead code which was originally for comparing OCR2 configs since this can be done in Gauntlet easily
3. addressed some more comments

ferglor commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,270 @@
+	NodeAddress           []string       `json:"nodeAddress"`
+	OcrSigningAddress     []string       `json:"ocrSigningAddress"`
+	PayeeAddress          common.Address `json:"payeeAddress"`
+	PeerId                []string       `json:"peerId"`
```
Should the json annotation here be `peerID` to be consistent with `ocr2ID`?

FelixFan1992 replied:
in fact, according to rddtool style, this should be `ocr2Id`. so i am working on a fix to clean up those before i fix this.

FelixFan1992 replied:
done

ferglor commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,270 @@
+	nodeInfos := h.fetchNodeInfos(ctx, log)
+	cnt := 0
+	for _, ni := range nodeInfos {
+		nodeAddr := ni.NodeAddress[0]
```
Should this be wrapped in a check that there are 1 or more elements in the `NodeAddress` slice and continue otherwise?

ferglor replied:
Also is there a particular significance to the zeroth Node Address? If so I would maybe maybe a function on the node receiver to fetch this?

ferglor commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,270 @@
+}
+
+func (h *baseHandler) fetchNodeInfos(ctx context.Context, log logger.Logger) []NodeInfo {
+	client := http.DefaultClient
```
I think the Defaul Client doesn't specify a request timeout, it might be better to instantiate a http Client and then specify a request timeout of like 10seconds or something?

FelixFan1992 replied:
done.
setting a timeout for 1 min just in case the weiwatcher URL has lots of content

ferglor commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,270 @@
+	defer wg.Done()
+
+	// get node addresses
+	resp, err := nodeRequest(cl, "/v2/keys/eth")
```
Can we extract the `/v2/keys/eth`, `/v2/keys/csa` and `/v2/keys/p2p` strings into constants? That way we can use the constant name to better describe what those URLs point to, e.g. `nodeAddressesEndpoint`, and then that would enable us to get rid of the comment above?

FelixFan1992 replied:
done

ferglor commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		log.Fatalf("failed to get node OCR2 config: %s", err)
+	}
+	ni.Ocr2ID = []string{ocr2Config.ID}
```
Can we add a function to the node info type that accepts the ocr2Config as a param, and then performs these operations within the function?

ferglor replied:
Or would it make sense to first perform all three requests, and then update the NodeInfo instance only if all three requests are successful? That way you can avoid partial updates on the NodeInfo, e.g. if the second or third request fails?

FelixFan1992 replied:
done

ferglor commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,270 @@
+	// CSA key for each chain. but this is still pending so assume only 1 CSA key on a node for now.
+	csaKey := &CSAKeyInfo{
+		NodeAddress: nodeAddresses[0],
+		PublicKey:   csaKeys[0].PubKey,
```
Should we add checks here to ensure there are one or more elements in each slice?

FelixFan1992 replied:
done

ferglor commented on core/scripts/chaincli/handler/scrape_node_config.go:
```diff
@@ -0,0 +1,283 @@
+	log.Warn("This does NOT scrape for payee address, admin address etc. Please verify that manually.")
+	cls := make([]cmd.HTTPClient, len(h.cfg.KeeperURLs))
+	for i := range h.cfg.KeeperURLs {
+		url := h.cfg.KeeperURLs[i]
```
No big deal, but you could get rid of this line if you changed the loop above to be:

```
for i, url := range h.cfg.KeeperURLs {
```

ferglor reviewed (approved):
Groovy! 👍
